libeos-payg/provider-loader.c

```diff
@@ -0,0 +1,310 @@
+{
+  g_return_if_fail (EPG_IS_PROVIDER_LOADER (self));
+  g_return_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable));
+
+  g_autoptr(GTask) task = g_task_new (self, cancellable, callback, user_data);
+  g_task_set_source_tag (task, epg_provider_loader_load_async);
+  g_task_set_task_data (task,
+                        g_ptr_array_new_with_free_func (g_object_unref),
+                        (GDestroyNotify) g_ptr_array_unref);
+  epg_multi_task_attach (task, 1);
+
+  GList *plugins = peas_engine_get_plugin_list (self->engine);
+  for (; plugins != NULL; plugins = plugins->next)
+    try_load (self, PEAS_PLUGIN_INFO (plugins->data), task);
+
+  maybe_return (task, NULL);
+}
+
+static void
+try_load (EpgProviderLoader *self,
+          PeasPluginInfo    *plugin_info,
+          GTask             *task)
+{
+  const gchar *name = peas_plugin_info_get_name (plugin_info);
+  g_autoptr(PeasExtension) extension = NULL;
+
+  if (!peas_engine_load_plugin (self->engine, plugin_info))
+    {
+      g_autoptr(GError) local_error = NULL;
+      gboolean ret = peas_plugin_info_is_available (plugin_info,
+                                                    &local_error);
+      if (ret)
+        {
+          /* Should not happen: peas_engine_load_plugin() should cause the
+           * PeasPluginInfo's error to be set.
+           */
+          g_warning ("Failed to load plugin %s", name);
+        }
+      else
+        {
+          g_warning ("Failed to load plugin %s: %s",
+                     name, local_error->message);
+        }
+
+      return;
+    }
+
+  if (!peas_engine_provides_extension (self->engine,
+                                       plugin_info,
+                                       EPG_TYPE_PROVIDER))
+    return;
+
+  extension = peas_engine_create_extension (self->engine,
+                                            plugin_info,
+                                            EPG_TYPE_PROVIDER,
+                                            NULL);
+  if (extension == NULL)
+    {
+      /* Should not happen: we just checked that plugin_info can do this. */
+      g_warning ("Failed to create EpgProvider from %s", name);
+      return;
+    }
+
+  if (!G_IS_ASYNC_INITABLE (extension))
+    {
+      g_warning ("EpgProvider from %s is not GAsyncInitable", name);
+      return;
+    }
+
+  epg_multi_task_increment (task);
+  g_async_initable_init_async (G_ASYNC_INITABLE (g_steal_pointer (&extension)),
+                               g_task_get_priority (task),
+                               g_task_get_cancellable (task),
+                               provider_init_cb,
+                               g_object_ref (task));
+}
+
+static void
+provider_init_cb (GObject      *source_object,
+                  GAsyncResult *result,
+                  gpointer      user_data)
+{
+  g_autoptr(EpgProvider) provider = EPG_PROVIDER (source_object);
+  g_autoptr(GTask) task = G_TASK (user_data);
+  g_autoptr(GError) error = NULL;
```

Comment:
Nitpick: I prefer the name `local_error` to be used for local `GError*` variables, to distinguish it from `GError **error` arguments.